src/Storage/Mysql/UserRepository.php

```diff
@@ -0,0 +1,39 @@
+<?php
+
+namespace Bugcache\Storage\Mysql;
+
+use Amp\Mysql\Pool;
+use Amp\Mysql\ResultSet;
+use Amp\Promise;
+use Bugcache\Storage;
+use Generator;
+use function Amp\resolve;
+
+class UserRepository implements Storage\UserRepository {
+    private $mysql;
+
+    public function __construct(Pool $mysql) {
+        $this->mysql = $mysql;
+    }
+
+    public function findByName(string $username): Promise {
+        $sql = "SELECT id, name FROM users WHERE name = ?";
+
+        return resolve($this->fetchUser($sql, [$username]));
+    }
+
+    private function fetchUser(string $sql, array $params = []): Generator {
+        /** @var ResultSet $result */
+        $result = yield $this->mysql->prepare($sql, $params);
```

Comment:
Amp is something different. There's no way to make it interoperable / swapable without the event loop standard.

We might want to use these classes in maintenance scripts or something like that that's not related to Aerys.

I could add something like the following method and cache the prepares, but do we really need that? Can't the pool itself cache something like that?

```php
    private function query(string $sql, array $params): Promise {
        if (isset($this->prepares[$sql])) {
            return $this->prepares[$sql]->execute($params);
        } else {
            $deferred = new Deferred;

            $this->mysql->prepare($sql)->when(function(Throwable $error = null, Mysql\Stmt $stmt = null) use ($deferred, $sql, $params) {
                if ($error) {
                    $deferred->fail($error);
                } else {
                    $this->prepares[$sql] = $stmt;
                    $deferred->succeed($stmt->execute($params));
                }
            });

            return $deferred->promise();
        }
    }
```

> The use case where they are executed multiple times at the same time should work flawlessly?

If I prepare a single time and execute twice or more in parallel, a new connection with a new prepare has to be opened anyway? Where's the gain of the prepare beforehand then?